### ISSUE TITLE
mouse wheel scroll events

### ISSUE THREAD

**Issue by scandum** [mintty/mintty#1032]
When the Alternate Scroll mode is enabled the scrollwheel sends 3 arrow up/down sequences per turn.

**Comment by mintty** [comment]
Yes, that is on purpose. The value is taken from a system setting (SPI_GETWHEELSCROLLLINES).
Use Control for single wheel events.

**Comment by mintty** [comment]
Not depending on alternate screen buffer, by the way; changed issue title.

**Comment by mintty** [comment]
Actually, Ctrl+wheel only works if corresponding font zooming is disabled. To be considered.

**Comment by scandum** [comment]
The issue specifically relates to Alternate Scroll mode.  From the xterm documentation:

> By default, the wheel mouse events (buttons 4 and 5) are translated to
> scroll-back and scroll-forw actions, respectively.  Those actions nor-
> mally scroll the whole window, as if the scrollbar was used.
> 
> However if Alternate Scroll mode is set, then cursor up/down controls
> are sent when the terminal is displaying the Alternate Screen Buffer.
> The initial state of Alternate Scroll mode is set using the alternate-
> Scroll resource.

When you're in bash in alternate scroll mode it'll scroll the input history up/down by 3 items per event, while holding shift will send 3 page-up/down requests. 

While the behavior is sensible for normal scroll mode, it isn't for the alternate mode.

Nor can I easily fix it client side because there's no way to distinguish between the scroll wheel and the keyboard in this mode since they send the same sequence.

**Comment by mintty** [comment]
I meant the 3-notches count applies to all mouse modes, not just alternate screen buffer virtual up/down scrolling.
I wouldn't agree this should be distinguished but that's a personal preference.
I am open to a proposal how to change things and stay maximal compatible...

**Comment by scandum** [comment]
Hard to predict compatibility, I honestly don't see a practical use case for anyone the way the alternative scroll mode behaves currently, so I think it's safe to change the behavior just for the alternative scroll mode and see if anyone complains.

**Comment by avih** [comment]
> When you're in bash in alternate scroll mode it'll scroll the input history up/down by 3 items per event, while holding shift will send 3 page-up/down requests.

> While the behavior is sensible for normal scroll mode, it isn't for the alternate mode.

Are you mixing between the terminal's alternate screen (e.g. when running `vim`) to some "alternate-scroll" mode of the terminal?
If you're NOT talking about "alternate screen" or "alternate buffer", then I don't know if mintty has the same "alternate scroll" mode which xterm has.

When the current running application does not handle mouse wheel on its own - then mintty applies useful some functionality of its own.

This mintty functionality is scrolling the backbuffer when at the main screen, and sending arrows (or maybe ^E/^Y) in alternate screen/buffer (NOT alternate scroll "mode").

The number of times it performs this action is however you configured your mouse wheel at the windows control panel. If you want it to send just one scroll thing then you should change it at the Windows control panel.

Alternatively, if @mintty wants to support some custom number of events to send regardless of the Windows configuration, then it's up to him.

**Comment by mintty** [comment]
Proposal: config file option `LinesPerMouseWheelNotch` would override the system factor if set, limited by terminal height.

**Comment by mintty** [comment]
I've uploaded a patch so that in fact (as I had incorrectly stated already above) Ctrl+mouse-wheel scrolls by only 1 line, if setting `ZoomMouse=false` (as otherwise Ctrl+wheel would zoom the font size).
Maybe that's sufficient and we can do without lines-per-notch configuration?

**Comment by scandum** [comment]
A LinesPerMouseWheelNotch config option would be useless for my purposes, as previously explained, so no need to bother on my account.

**Comment by avih** [comment]
> When the Alternate Scroll mode is enabled the scrollwheel sends 3 arrow up/down sequences per turn.

@scandum can you please give some details of this "alternate scroll mode"?

Specifically, when I scroll the mouse wheel in bash - it just scrolls into the backbuffer, it doesn't send any arrow keys.

How can one enable this "alternate scroll mode"? and how to reproduce the behavior you're describing?

**Comment by scandum** [comment]
@avih It's in the manual: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html

**Comment by avih** [comment]
@scandum but this is mintty, not xterm, and in `man mintty` I can't find it mentioned. So how do you enable this mode in mintty?

**Comment by mintty** [comment]
https://github.com/mintty/mintty/wiki/CtrlSeqs#mousewheel-reporting

**Comment by mintty** [comment]
@scandum,

> LinesPerMouseWheelNotch config option would be useless for my purposes, as previously explained

not sure where you explained that. You could set the value to 1, isn't that what you want to achieve?

**Comment by avih** [comment]
Well, in `man xterm` it's described like so:

```
alternateScroll (class ScrollCond)
    If “true”, the scroll-back and scroll-forw actions send cursor-up and
    -down keys when xterm is displaying the alternate screen.  The default
    is “false”.

    The alternateScroll state can also be set using a control sequence.
```

So it appears to be a mouse-wheel mode, possible enabled by default, which only has an effect in alt-screen.

As far as I can tell this send up/down arrows, or maybe ^Y/^E, for instance inside `nano` or `vim` when/if mouse-wheel is not enabled by those applications.

However, I don't think bash uses the alternate screen, at least not by default, so I don't yet get how it applies to bash - as described initially. When I scroll in bash then it just scrolls the scrollbar, or does nothing if there's no scrollbar yet (e.g. right after I open a mintty window).

Regardless, the number of events which mintty sends in this case can be configured from the windows control panel, and @mintty also suggested to add an option to override it with a custom value - but I don't understand why @scandum said it would not help his case.

**Comment by BrianInglis** [comment]
Check out above xterm link under `CSI ? Pm h` `Ps = 1 0 0 7` Enable `CSI ? Pm l` `Ps = 1 0 0 7` "Disable Alternate Scroll Mode, `xterm`. This corresponds to the _alternateScroll_ resource." -> 
"If *true*, the _scroll-back_ and _scroll-forw_ actions send `cursor-up` and `-down` keys when `xterm` is displaying the alternate screen. The default is *false*. The _alternateScroll_ state can also be set using a control sequence."
So, similar to _Application Keypad_ mode, _Alternate Scroll_ mode sends scroll actions as cursor key movements to the application using the alternate screen buffer, so it works with `curses` apps.

**Comment by mintty** [comment]
> I don't think bash uses the alternate screen, at least not by default, 

Ah, right, so mentioning bash as an example in the mintty manual is confusing, I'll remove it. (Although you can of course run bash in alternate screen mode.)

**Comment by avih** [comment]
> Although you can of course run bash in alternate screen mode

Maybe, but I couldn't find it mentioned at the bash manual on cygwin (4.4.12) or Ubuntu 18.04 (4.4.20). Or do you mean to send an escape sequence to enter alt-screen manually?

Additionally, why would one even want to run bash in alt-screen? you lose the scrollback buffer and you gain.. what exactly?

**Comment by mintty** [comment]
Added two features:
* option LinesPerMouseWheelNotch
* enforce mousewheel reporting with Alt, so shell history can be scrolled with Control+Alt+mouse-wheel

**Comment by scandum** [comment]
Nice work on the documentation. Still a poor choice of behavior, but it is what it is, and with 'less' using it that's probably how it'll stay until the end of times.

The \e[?7786h \e[?7787h combination is useful and I think I can work with that.

**Comment by mintty** [comment]
Released 3.4.0.
